cudax/include/cuda/experimental/__container/vector.h

```diff
@@ -0,0 +1,1252 @@
+#include <cuda/experimental/__detail/utility.cuh>
+#include <cuda/experimental/__utility/select_execution_space.h>
+
+#if _CCCL_STD_VER >= 2014 && !defined(_CCCL_COMPILER_MSVC_2017) \
+  && defined(LIBCUDACXX_ENABLE_EXPERIMENTAL_MEMORY_RESOURCE)
+
+//! @file The \c vector class provides a container of contiguous memory
+namespace cuda::experimental
+{
+
+//! @rst
+//! .. _cudax-containers-vector:
+//!
+//! vector
+//! -------
+//!
+//! ``vector`` is a container that provides resizable typed storage allocated from a given :ref:`memory resource
+//! <libcudacxx-extended-api-memory-resources-resource>`. It handles alignment, release and growth of the allocation.
+//! The elements are initialized during construction, if the storage is not user provided through a
+//! :ref:`uninitialized_buffer <cudax-containers-uninitialized-buffer>` with matching properties.
+//!
+//! In addition to being type-safe, ``vector`` also takes a set of :ref:`properties
+//! <libcudacxx-extended-api-memory-resources-properties>` to ensure that e.g. execution space constraints are checked
+//! at compile time. However, only stateless properties can be forwarded. To use a stateful property,
+//! implement :ref:`get_property(const vector&, Property) <libcudacxx-extended-api-memory-resources-properties>`.
+//!
+//! .. warning::
+//!
+//!    ``vector`` stores a reference to the provided memory :ref:`memory resource
+//!    <libcudacxx-extended-api-memory-resources-resource>`. It is the user's responsibility to ensure the lifetime of
+//!    the resource exceeds the lifetime of the vector.
+//!
+//! @endrst
+//! @tparam _Tp the type to be stored in the buffer
+//! @tparam _Properties... The properties the allocated memory satisfies
+template <class _Tp, class... _Properties>
+class vector
+{
+public:
+  using value_type             = _Tp;
+  using reference              = _Tp&;
+  using const_reference        = const _Tp&;
+  using pointer                = _Tp*;
+  using const_pointer          = const _Tp*;
+  using iterator               = heterogeneous_iterator<_Tp, false, _Properties...>;
+  using const_iterator         = heterogeneous_iterator<_Tp, true, _Properties...>;
+  using reverse_iterator       = _CUDA_VSTD::reverse_iterator<iterator>;
+  using const_reverse_iterator = _CUDA_VSTD::reverse_iterator<const_iterator>;
+  using size_type              = _CUDA_VSTD::size_t;
+  using __resource_ref         = _CUDA_VMR::resource_ref<_Properties...>;
+
+  // Doxygen cannot handle the macro expansions
+  template <class _Range>
+  static constexpr bool __compatible_range = _CUDA_VRANGES::__container_compatible_range<_Range, _Tp>;
+
+private:
+  uninitialized_buffer<_Tp, _Properties...> __buf_;
+  size_type __size_ = 0; // initialized to 0 in case initialization of the elements might throw
+
+  //! @brief In can an exception is thrown adjusts the size of the vector so that the previously constructed elements
+  //! are accounted for
+  struct _Adopt_size
```

Comment:
I think you are misusing "adopt". It means "choose to take up, follow, or use".  Maybe "adjust", as used in the doc string, is better.

```suggestion
  struct _Adapt_size
```